### PR TITLE
Can implementation; DemoCan project with Raspi platform impl

### PR DESCRIPTION
**Pull request by samparent97**

### CODE REVIEW

File: firmware/CMakeLists.txt [hunk 1]
```diff
@@ -63,9 +63,11 @@ add_library(bindings)
 add_subdirectory(${DIR_PROJECT})  # modifies "main" (executable)
 add_subdirectory(${DIR_PLATFORM})  # modifies "bindings" (library)
 
+set(DIR_ETL "${CMAKE_CURRENT_SOURCE_DIR}/third-party/etl/include")
```
Comment (BlakeFreer) [inline firmware/CMakeLists.txt:66]: `add_subirectory(${CUR_SRC}.../etl')

File: scripts/cangen/can_gen.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                break
+
+        self.logger.info("Little endian mask generated")
+
+        return mask, start
+
+    def __big_endian_shift_amounts(self, num_zeroes: int) -> List[int]:
+        shift_amounts = [0] * self.EIGHT_BYTES
+        for i in range(0, self.EIGHT_BYTES):
+            shift_amounts[-i] = (i * self.EIGHT_BITS) - num_zeroes
```
[the remark this answers is not in the document]
Comment (BlakeFreer) [reply scripts/cangen/can_gen.py:154]: replace with 
`return list(range(-num_zeros, 64-num_zeros, 8))[::-1]`

**Review by BlakeFreer — CHANGES_REQUESTED** [review]

**Comment by BlakeFreer** [comment]
rebase

File: firmware/mcal/raspi/periph/can.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                can_queue_.pop();
+                msg_idx++;
+            }
+        }
+    }
+
+private:
+    using RawCanMsg = shared::can::RawCanMsg;
+
+    static constexpr uint8_t kMaxMsgBytes = 8;
```
[the remark this answers is not in the document]
Comment (samparent97) [reply firmware/mcal/raspi/periph/can.h:84]: yes

File: firmware/mcal/raspi/periph/can.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        if (bind(sock_, (struct sockaddr*)&sock_addr_, sizeof(sock_addr_)) <
+            0) {
+            perror("Error binding socket");
+            close(sock_);
+        }
+
+        reader_thread = std::thread(&CanBase::StartReading, this);
+    }
+
+    void Send(const shared::can::RawCanMsg& can_tx_msg) {
```
[the remark this answers is not in the document]
Comment (samparent97) [reply firmware/mcal/raspi/periph/can.h:57]: removed the using

File: firmware/projects/DemoCan/generated/can_messages.h [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        wheel_speed = (static_cast<float>(temp_wheel_speed) * kWheelSpeedScale) + kWheelSpeedOffset;
+
+    }
+};
+
+/* -------------------- TX MESSAGES -------------------- */
+
+class TempSensors : public shared::can::CanTxMsg {
+public:
+    float sensor6 = 0;
```
[the remark this answers is not in the document]
Comment (samparent97) [reply firmware/projects/DemoCan/generated/can_messages.h]: demo can has them in this reversed order by the bits in the raw message

File: firmware/projects/DemoCan/generated/msg_registry.h
[the remark this answers is not in the document]
Comment (samparent97) [reply firmware/projects/DemoCan/generated/msg_registry.h:1]: added #102

File: firmware/projects/DemoCan/main.cc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+int main(void) {
+    bindings::Initialize();
+    std::chrono::milliseconds duration(100);
+
+    generated::can::TempSensors temp_sens_msg;
+    generated::can::VehicleInfo veh_info_msg;
+
+    int i = 0;
+    while (1) {
```
[the remark this answers is not in the document]
Comment (samparent97) [reply firmware/projects/DemoCan/main.cc:32]:
```suggestion
    while (true) {
```

**Pull request merged**